File: backend/agent_executor.ts
/**
 * agent_executor.ts — Valentins AI Agent Backend
 *
 * Builds and executes Programmable Transaction Blocks (PTBs)
 * against the quantum_vault smart contracts on Sui.
 *
 * Install:
 *   npm install @mysten/sui @mysten/bcs dotenv
 *
 * Usage:
 *   AGENT_PRIVATE_KEY=suiprivkey1... npx ts-node agent_executor.ts
 */

import {
  SuiClient,
  getFullnodeUrl,
} from '@mysten/sui/client';
import { Transaction } from '@mysten/sui/transactions';
import { Ed25519Keypair } from '@mysten/sui/keypairs/ed25519';
import { fromBase64 } from '@mysten/sui/utils';
import 'dotenv/config';

// ═══════════════════════════════════════════════════════════
//  CONFIG
// ═══════════════════════════════════════════════════════════

const NETWORK = (process.env.SUI_NETWORK as 'devnet' | 'testnet' | 'mainnet') ?? 'devnet';
const RPC_URL = process.env.SUI_RPC_URL ?? getFullnodeUrl(NETWORK);

// On-chain IDs — filled after deploy.sh
const PACKAGE_ID    = process.env.PACKAGE_ID!;
const PORTFOLIO_ID  = process.env.PORTFOLIO_ID!;
const AGENT_CAP_ID  = process.env.AGENT_CAP_ID!;
const SUI_CLOCK     = '0x6';

// ═══════════════════════════════════════════════════════════
//  KEYPAIR
// ═══════════════════════════════════════════════════════════

function loadKeypair(): Ed25519Keypair {
  const key = process.env.AGENT_PRIVATE_KEY;
  if (!key) throw new Error('AGENT_PRIVATE_KEY not set');

  // Support both raw hex and Sui-encoded (suiprivkey1...) formats
  if (key.startsWith('suiprivkey')) {
    return Ed25519Keypair.fromSecretKey(key);
  }
  // Raw base64 or hex
  const bytes = key.startsWith('0x')
    ? Uint8Array.from(Buffer.from(key.slice(2), 'hex'))
    : fromBase64(key);
  return Ed25519Keypair.fromSecretKey(bytes);
}

// ═══════════════════════════════════════════════════════════
//  CLIENT
// ═══════════════════════════════════════════════════════════

const client = new SuiClient({ url: RPC_URL });
const keypair = loadKeypair();

console.log(`🔑 Agent address: ${keypair.getPublicKey().toSuiAddress()}`);
console.log(`🌐 Network: ${NETWORK} (${RPC_URL})`);
console.log(`📦 Package: ${PACKAGE_ID}`);

// ═══════════════════════════════════════════════════════════
//  PTB BUILDERS
// ═══════════════════════════════════════════════════════════

/**
 * Demo trade — calls execute_rebalance (no fund movement).
 * Perfect for the pitch: shows guardrails blocking / allowing.
 */
export async function executeRebalance(
  amountMist: number,
  isQuantumOptimized: boolean = true,
): Promise<string> {
  const tx = new Transaction();

  tx.moveCall({
    target: `${PACKAGE_ID}::portfolio::execute_rebalance`,
    arguments: [
      tx.object(AGENT_CAP_ID),
      tx.object(PORTFOLIO_ID),
      tx.pure.u64(amountMist),
      tx.pure.bool(isQuantumOptimized),
      tx.object(SUI_CLOCK),
    ],
  });

  const result = await client.signAndExecuteTransaction({
    signer: keypair,
    transaction: tx,
    options: { showEffects: true, showEvents: true },
  });

  console.log('✅ TX digest:', result.digest);
  if (result.events) {
    for (const ev of result.events) {
      console.log(`  📢 ${ev.type}`, ev.parsedJson);
    }
  }
  return result.digest;
}

/**
 * Production PTB — withdraw → DEX swap → deposit back.
 *
 *   Step 1: withdraw_for_swap  → Coin<SUI>
 *   Step 2: (DEX call — placeholder, e.g. Cetus)
 *   Step 3: deposit_returns    ← Coin<SUI> back to vault
 *
 * All three calls happen in ONE atomic transaction.
 * If any guardrail fails, the ENTIRE PTB reverts.
 */
export async function swapRebalance(
  amountMist: number,
  minOutput: number = 0,
  isQuantumOptimized: boolean = true,
): Promise<string> {
  const tx = new Transaction();

  // Step 1: withdraw SUI from the vault
  const [coin] = tx.moveCall({
    target: `${PACKAGE_ID}::portfolio::withdraw_for_swap`,
    arguments: [
      tx.object(AGENT_CAP_ID),
      tx.object(PORTFOLIO_ID),
      tx.pure.u64(amountMist),
      tx.pure.bool(isQuantumOptimized),
      tx.object(SUI_CLOCK),
    ],
  });

  // Step 2: DEX swap — PLACEHOLDER
  // In production, replace with Cetus / Aftermath / DeepBook call:
  //
  //   const [outputCoin] = tx.moveCall({
  //     target: `${CETUS_PACKAGE}::router::swap_exact_input`,
  //     arguments: [coin, ...],
  //   });
  //
  // For now, we deposit the same coin back (round-trip demo):
  const outputCoin = coin;

  // Step 3: deposit the swap result back into the vault
  tx.moveCall({
    target: `${PACKAGE_ID}::portfolio::deposit_returns`,
    arguments: [
      tx.object(PORTFOLIO_ID),
      outputCoin,
      tx.pure.u64(minOutput),       // slippage guardrail
    ],
  });

  const result = await client.signAndExecuteTransaction({
    signer: keypair,
    transaction: tx,
    options: { showEffects: true, showEvents: true },
  });

  console.log('✅ Swap TX digest:', result.digest);
  if (result.events) {
    for (const ev of result.events) {
      console.log(`  📢 ${ev.type}`, ev.parsedJson);
    }
  }
  return result.digest;
}

/**
 * Dry-run a transaction WITHOUT submitting to the chain.
 * Returns { success: boolean, error?: string }.
 */
export async function dryRun(
  amountMist: number,
  isQuantumOptimized: boolean = true,
): Promise<{ success: boolean; error?: string }> {
  const tx = new Transaction();

  tx.moveCall({
    target: `${PACKAGE_ID}::portfolio::execute_rebalance`,
    arguments: [
      tx.object(AGENT_CAP_ID),
      tx.object(PORTFOLIO_ID),
      tx.pure.u64(amountMist),
      tx.pure.bool(isQuantumOptimized),
      tx.object(SUI_CLOCK),
    ],
  });

  tx.setSender(keypair.getPublicKey().toSuiAddress());

  try {
    const dryResult = await client.dryRunTransactionBlock({
      transactionBlock: await tx.build({ client }),
    });

    const status = dryResult.effects.status;
    if (status.status === 'success') {
      return { success: true };
    }
    return { success: false, error: status.error ?? 'unknown error' };
  } catch (e: any) {
    return { success: false, error: e.message };
  }
}

// ═══════════════════════════════════════════════════════════
//  SWAP AND REBALANCE (atomic demo with slippage + quantum score)
// ═══════════════════════════════════════════════════════════

export async function swapAndRebalance(
  amountMist: number,
  minOutput: number,
  quantumScore: number = 0,
  isQuantumOptimized: boolean = true,
): Promise<string> {
  const tx = new Transaction();

  tx.moveCall({
    target: `${PACKAGE_ID}::portfolio::swap_and_rebalance`,
    arguments: [
      tx.object(AGENT_CAP_ID),
      tx.object(PORTFOLIO_ID),
      tx.pure.u64(amountMist),
      tx.pure.u64(minOutput),
      tx.pure.bool(isQuantumOptimized),
      tx.pure.u64(quantumScore),
      tx.object(SUI_CLOCK),
    ],
  });

  const result = await client.signAndExecuteTransaction({
    signer: keypair,
    transaction: tx,
    options: { showEffects: true, showEvents: true },
  });

  console.log('✅ swap_and_rebalance TX:', result.digest);
  if (result.events) {
    for (const ev of result.events) {
      console.log(`  📢 ${ev.type}`, ev.parsedJson);
    }
  }
  return result.digest;
}

// ═══════════════════════════════════════════════════════════
//  SET PAUSED  (Admin emergency kill-switch)
// ═══════════════════════════════════════════════════════════

const ADMIN_CAP_ID = process.env.ADMIN_CAP_ID!;

export async function setPaused(paused: boolean): Promise<string> {
  const tx = new Transaction();

  tx.moveCall({
    target: `${PACKAGE_ID}::portfolio::set_paused`,
    arguments: [
      tx.object(ADMIN_CAP_ID),
      tx.object(PORTFOLIO_ID),
      tx.pure.bool(paused),
    ],
  });

  const result = await client.signAndExecuteTransaction({
    signer: keypair,
    transaction: tx,
    options: { showEffects: true, showEvents: true },
  });

  console.log(`${paused ? '🛑 PAUSED' : '▶️  RESUMED'} TX: ${result.digest}`);
  return result.digest;
}

// ═══════════════════════════════════════════════════════════
//  EVENT STREAMER  (for Person C / Frontend dashboard)
// ═══════════════════════════════════════════════════════════

export async function streamTradeEvents() {
  console.log('📡 Subscribing to TradeEvent...');

  // Note: WebSocket subscriptions require a WS-enabled RPC
  const unsubscribe = await client.subscribeEvent({
    filter: {
      MoveEventType: `${PACKAGE_ID}::portfolio::TradeEvent`,
    },
    onMessage(event) {
      const d = event.parsedJson as any;
      console.log(`\n🔔 Trade #${d.trade_id}`);
      console.log(`   Agent:    ${d.agent_address}`);
      console.log(`   Amount:   ${d.input_amount} MIST`);
      console.log(`   Balance:  ${d.balance_before} → ${d.balance_after}`);
      console.log(`   Quantum:  ${d.is_quantum_optimized ? '⚛️  YES' : '❌ NO'}`);
    },
  });

  // Keep alive — Ctrl+C to stop
  process.on('SIGINT', async () => {
    await unsubscribe();
    process.exit(0);
  });
}

// ═══════════════════════════════════════════════════════════
//  CLI
// ═══════════════════════════════════════════════════════════

async function main() {
  const cmd = process.argv[2] ?? 'demo';
  const amount = parseInt(process.argv[3] ?? '1000000000', 10); // default 1 SUI

  switch (cmd) {
    case 'demo':
      console.log(`\n🚀 Demo rebalance: ${amount} MIST`);
      await executeRebalance(amount, true);
      break;

    case 'swap':
      console.log(`\n🔄 Swap rebalance: ${amount} MIST`);
      await swapRebalance(amount, 0, true);
      break;

    case 'quantum':
      console.log(`\n⚛️  Quantum swap_and_rebalance: ${amount} MIST`);
      const minOut = parseInt(process.argv[4] ?? String(Math.floor(amount * 0.9)), 10);
      const qScore = parseInt(process.argv[5] ?? '85', 10);
      await swapAndRebalance(amount, minOut, qScore, true);
      break;

    case 'pause':
      console.log('\n🛑 Pausing portfolio...');
      await setPaused(true);
      break;

    case 'resume':
      console.log('\n▶️  Resuming portfolio...');
      await setPaused(false);
      break;

    case 'dryrun':
      console.log(`\n🧪 Dry-run: ${amount} MIST`);
      const result = await dryRun(amount, true);
      if (result.success) {
        console.log('✅ Dry-run PASSED — guardrails OK');
      } else {
        console.log(`🛑 Dry-run BLOCKED: ${result.error}`);
      }
      break;

    case 'killswitch':
      // "Fat Finger" test: try to trade 100% of portfolio
      console.log('\n💀 KILL-SWITCH TEST: attempting 100% portfolio drain...');
      const portfolio = await client.getObject({
        id: PORTFOLIO_ID,
        options: { showContent: true },
      });
      const fields = (portfolio.data?.content as any)?.fields;
      const fullBalance = parseInt(fields?.balance ?? '999999999999', 10);
      console.log(`   Portfolio balance: ${fullBalance} MIST`);
      console.log(`   Attempting trade:  ${fullBalance} MIST (100%)`);

      const ks = await dryRun(fullBalance, false);
      if (ks.success) {
        console.log('⚠️  Trade would PASS — check your guardrails!');
      } else {
        console.log(`🛡️  BLOCKED by guardrails: ${ks.error}`);
        console.log('✅ Kill-switch working! Portfolio is safe.');
      }
      break;

    case 'stream':
      await streamTradeEvents();
      break;

    default:
      console.log('Usage: npx ts-node agent_executor.ts <demo|swap|quantum|dryrun|killswitch|pause|resume|stream> [amount_mist] [min_output] [quantum_score]');
  }
}

main().catch(console.error);

File: backend/blockchain/agent_executor.py
#!/usr/bin/env python3
"""
agent_executor.py — AI Agent Backend CLI

Builds and executes Programmable Transaction Blocks (PTBs)
against the quantum_vault smart contracts on Sui.

Usage:
    python agent_executor.py demo [amount_mist]
    python agent_executor.py swap [amount_mist]
    python agent_executor.py quantum [amount_mist] [min_output] [quantum_score]
    python agent_executor.py dryrun [amount_mist]
    python agent_executor.py killswitch
    python agent_executor.py pause
    python agent_executor.py resume
    python agent_executor.py stream
"""

import asyncio
import json
import logging
import os
import sys
import time
from typing import Any, Dict

import httpx
from dotenv import load_dotenv

from core.error_map import parse_abort_error
from blockchain.ptb_builder import (
    TxResult,
    _run_sui_cmd,
    build_execute_rebalance,
    build_set_paused,
    build_swap_and_rebalance,
    dry_run_rebalance,
    dry_run_swap,
    execute_rebalance,
    execute_set_paused,
    execute_swap_and_rebalance,
)

load_dotenv()

logging.basicConfig(level=logging.INFO, format="%(message)s")
logger = logging.getLogger(__name__)

# ═══════════════════════════════════════════════════════════
#  CONFIG
# ═══════════════════════════════════════════════════════════

NETWORK = os.getenv("SUI_NETWORK", "devnet")
RPC_URL = os.getenv("SUI_RPC_URL", f"https://fullnode.{NETWORK}.sui.io:443")
PACKAGE_ID = os.getenv("PACKAGE_ID", "")
PORTFOLIO_ID = os.getenv("PORTFOLIO_ID", os.getenv("PORTFOLIO_OBJECT_ID", ""))
AGENT_CAP_ID = os.getenv("AGENT_CAP_ID", "")

_rpc_id = 0


def _rpc_call(method: str, params: list) -> dict:
    global _rpc_id
    _rpc_id += 1
    payload = {"jsonrpc": "2.0", "id": _rpc_id, "method": method, "params": params}
    with httpx.Client(timeout=15) as client:
        resp = client.post(RPC_URL, json=payload)
        resp.raise_for_status()
        data = resp.json()
    if "error" in data:
        raise RuntimeError(f"RPC error: {data['error']}")
    return data.get("result", {})


def _get_portfolio_balance() -> int:
    """Fetch current portfolio balance in MIST."""
    obj = _rpc_call(
        "sui_getObject",
        [
            PORTFOLIO_ID,
            {"showContent": True},
        ],
    )
    fields = obj.get("data", {}).get("content", {}).get("fields", {})
    return int(fields.get("balance", "0"))


# ═══════════════════════════════════════════════════════════
#  COMMANDS
# ═══════════════════════════════════════════════════════════


def cmd_demo(amount: int):
    """Demo rebalance — calls execute_rebalance (no fund movement)."""
    print(f"\n Demo rebalance: {amount} MIST")
    result = execute_rebalance(amount, True)
    if result.success:
        print(f" TX digest: {result.digest}")
        for ev in result.events:
            print(f"   {ev['type']}: {json.dumps(ev['data'], indent=2)}")
    else:
        print(f" Failed: {result.error}")


def cmd_swap(amount: int):
    """Swap rebalance — withdraw → DEX swap → deposit back."""
    print(f"\n Swap rebalance: {amount} MIST")
    result = execute_swap_and_rebalance(amount, 0, True, 0)
    if result.success:
        print(f" Swap TX digest: {result.digest}")
        for ev in result.events:
            print(f"   {ev['type']}: {json.dumps(ev['data'], indent=2)}")
    else:
        print(f" Failed: {result.error}")


def cmd_quantum(amount: int, min_output: int, quantum_score: int):
    """Quantum swap_and_rebalance with score."""
    print(f"\n  Quantum swap_and_rebalance: {amount} MIST")
    result = execute_swap_and_rebalance(amount, min_output, True, quantum_score)
    if result.success:
        print(f" Quantum TX: {result.digest}")
        for ev in result.events:
            print(f"   {ev['type']}: {json.dumps(ev['data'], indent=2)}")
    else:
        print(f" Failed: {result.error}")


def cmd_dryrun(amount: int):
    """Dry-run a transaction WITHOUT submitting."""
    print(f"\n Dry-run: {amount} MIST")
    result = dry_run_rebalance(amount, True)
    if result.success:
        print(" Dry-run PASSED — guardrails OK")
    else:
        parsed = parse_abort_error(result.error)
        print(f" Dry-run BLOCKED: {parsed.frontend_message}")


def cmd_killswitch():
    """Fat Finger test: try to trade 100% of portfolio."""
    print("\n KILL-SWITCH TEST: attempting 100% portfolio drain...")
    try:
        balance = _get_portfolio_balance()
        print(f"   Portfolio balance: {balance} MIST")
        print(f"   Attempting trade:  {balance} MIST (100%)")

        result = dry_run_rebalance(balance, False)
        if result.success:
            print("  Trade would PASS — check your guardrails!")
        else:
            parsed = parse_abort_error(result.error)
            print(f"  BLOCKED by guardrails: {parsed.frontend_message}")
            print(" Kill-switch working! Portfolio is safe.")
    except Exception as e:
        print(f" Error: {e}")


def cmd_pause():
    """Emergency pause."""
    print("\n Pausing portfolio...")
    result = execute_set_paused(True)
    if result.success:
        print(f" PAUSED — TX: {result.digest}")
    else:
        print(f" Failed: {result.error}")


def cmd_resume():
    """Resume from pause."""
    print("\n▶  Resuming portfolio...")
    result = execute_set_paused(False)
    if result.success:
        print(f"▶  RESUMED — TX: {result.digest}")
    else:
        print(f" Failed: {result.error}")


def cmd_stream():
    """Stream TradeEvents from the chain (poll-based)."""
    print(" Streaming TradeEvent...")
    print("   Press Ctrl+C to stop\n")

    event_type = f"{PACKAGE_ID}::portfolio::TradeEvent"
    seen = set()

    try:
        while True:
            try:
                result = _rpc_call(
                    "suix_queryEvents",
                    [
                        {"MoveEventType": event_type},
                        None,
                        5,
                        True,
                    ],
                )
                events = result.get("data", [])
                for ev in events:
                    ev_id = ev.get("id", {})
                    key = f"{ev_id.get('txDigest')}:{ev_id.get('eventSeq')}"
                    if key in seen:
                        continue
                    seen.add(key)

                    d = ev.get("parsedJson", {})
                    print(f"\n Trade #{d.get('trade_id', '?')}")
                    print(f"   Agent:    {d.get('agent_address', '?')}")
                    print(f"   Amount:   {d.get('input_amount', '?')} MIST")
                    print(
                        f"   Balance:  {d.get('balance_before', '?')} → {d.get('balance_after', '?')}"
                    )
                    is_q = d.get("is_quantum_optimized", False)
                    print(f"   Quantum:  {'  YES' if is_q else ' NO'}")
            except Exception:
                pass

            time.sleep(3)
    except KeyboardInterrupt:
        print("\n Stopped.")


# ═══════════════════════════════════════════════════════════
#  CLI
# ═══════════════════════════════════════════════════════════


def main():
    print(f" Network: {NETWORK} ({RPC_URL})")
    print(f" Package: {PACKAGE_ID}")

    cmd = sys.argv[1] if len(sys.argv) > 1 else "demo"
    amount = int(sys.argv[2]) if len(sys.argv) > 2 else 1_000_000_000  # default 1 SUI

    if cmd == "demo":
        cmd_demo(amount)
    elif cmd == "swap":
        cmd_swap(amount)
    elif cmd == "quantum":
        min_out = int(sys.argv[3]) if len(sys.argv) > 3 else int(amount * 0.9)
        q_score = int(sys.argv[4]) if len(sys.argv) > 4 else 85
        cmd_quantum(amount, min_out, q_score)
    elif cmd == "pause":
        cmd_pause()
    elif cmd == "resume":
        cmd_resume()
    elif cmd == "dryrun":
        cmd_dryrun(amount)
    elif cmd == "killswitch":
        cmd_killswitch()
    elif cmd == "stream":
        cmd_stream()
    else:
        print(
            "Usage: python agent_executor.py "
            "<demo|swap|quantum|dryrun|killswitch|pause|resume|stream> "
            "[amount_mist] [min_output] [quantum_score]"
        )


if __name__ == "__main__":
    main()
